_posts/2014-07-05-logging-in-python.md

```diff
@@ -169,6 +182,38 @@ handlers:
 
 Notice how `filters` is an array. Therefore we can add as many filters as we need here.
 
+## Redacting logs using a Custom Formatter
+
+In this approach, we will use a custom formatter objects that simply wrap the real formatter objects such that, messages can be intercepted right after they have been formatted but before the `handler` gets a chance to write them. This has the advantage that it is very simple, but unlike the previous approach, it can only be configured programmatically (which is actually good in a way, as it will always get applied irrespective of what the user configures).
+
+First we will create a formatter class which is just a wrapper that simply *proxies* all the original formatter attributes, except for the `format()` method.
+
+{% highlight python linenos %}
+
+class RedactingFormatter(object):
+    def __init__(self, orig_filter, patterns):
+        self.orig_filter = orig_filter
+        self._patterns = patterns
+
+    def format(self, record):
+        msg = self.orig_filter.format(record)
+        for pattern in self._patterns:
+            msg = msg.replace(pattern, "***")
+        return msg
+
+    def __getattr__(self, attr):
+        return getattr(self.orig_filter, attr)
+        
+{% endhighlight %}
+
+In the code right after `logging` is configured ([see](#configuring-your-loggers)), we need to add the below code to wrap all formatters (*a word of caution*: the below code assumes that all the handlers are on the root logger, but this may not always be the case):
+
+{% highlight python linenos %}
+for h in logging.root.handlers:
+    h.setFormatter(FilterFormatter(h.formatter))
+{% endhighlight %}
```

Comment:
I do not think this is possible. The keywords are expanded and wired by logging config and since there is no need to wire two formatters together, it doesn't come with a way to express this.